p2p/sender.go

```diff
@@ -26,18 +27,78 @@ import (
 
 	"github.com/obolnetwork/charon/app/errors"
 	"github.com/obolnetwork/charon/app/log"
+	"github.com/obolnetwork/charon/app/z"
 )
 
-// SendAsync sends a libp2p message and logs a warning on error.
-//   Usage: go p2p.SendAsync(ctx, tcpNode, protoId, peerId, msg)
-func SendAsync(ctx context.Context, tcpNode host.Host, protoID protocol.ID, peerID peer.ID, msg proto.Message) {
-	err := sendMsg(ctx, tcpNode, protoID, peerID, msg)
-	if err != nil {
-		log.Warn(ctx, "Failed sending p2p message", err)
+const (
+	senderHysteresis = 3
+	senderBuffer     = senderHysteresis + 1
+)
+
+type peerState struct {
+	failing bool
+	buffer  []error
+}
+
+// Sender provides an API for sending libp2p messages, both synchronous and asynchronous.
+// It also provides log filtering for async sending, mitigating
+// error storms when peers are down.
+type Sender struct {
+	states sync.Map // map[peer.ID]peerState
+}
+
+// addResult adds the result of sending a p2p message to the internal state and possibly logs a status change.
+func (s *Sender) addResult(ctx context.Context, peerID peer.ID, err error) {
+	var state peerState
+	if val, ok := s.states.Load(peerID); ok {
+		state = val.(peerState)
+	}
+
+	state.buffer = append(state.buffer, err)
+	if len(state.buffer) > senderBuffer { // Trim buffer
+		state.buffer = state.buffer[len(state.buffer)-senderBuffer:]
 	}
+
+	failure := err != nil
+	success := !failure
+
+	if success && state.failing {
+		// See if we have senderHysteresis successes i.o.t. change state to success.
+		full := len(state.buffer) == senderBuffer
+		oldestFailure := state.buffer[0] != nil
+		othersSuccess := true
+		for i := 1; i < len(state.buffer); i++ {
+			if state.buffer[i] != nil {
+				othersSuccess = false
+				break
+			}
+		}
+
+		if full && oldestFailure && othersSuccess {
+			state.failing = false
+			log.Info(ctx, "P2P sending recovered", z.Str("peer", PeerName(peerID)))
+		}
+	} else if failure && (len(state.buffer) == 1 || !state.failing) {
+		// First attempt failed or state changed to failing
+		log.Warn(ctx, "P2P sending failing", err, z.Str("peer", PeerName(peerID)))
+		state.failing = true
+	}
+
+	s.states.Store(peerID, state) // Note there is a race if two results for the same peer is added at the same time, but this isn't critical.
+}
+
+// SendAsync returns nil sends a libp2p message asynchronously. It logs results on state change (success to/from failure).
+func (s *Sender) SendAsync(ctx context.Context, tcpNode host.Host, protoID protocol.ID, peerID peer.ID, msg proto.Message) error {
```

Comment:
discussion: why does this method `SendAsync()` even return an error?

Reply:
so that it has the same API as `Send`, so components can use an abstract SendFunc and either use sync or async

Reply:
I'll move SendFunc here

Reply:
okay alright 👍